File: scripts/update_sanctions.py
#!/usr/bin/env python3
"""
scripts/update_sanctions.py — Download and refresh the OFAC SDN sanctions list.

Downloads the OFAC SDN Advanced XML, extracts all ETH addresses,
and stores them in the local DuckDB database.

Run this weekly to keep the list current.

Usage:
    python scripts/update_sanctions.py
"""

import sys
import asyncio
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from forensics.sanctions import update_sanctions_list, count_sanctioned


async def main():
    print("Chain Analytics — OFAC Sanctions Update")
    print("=" * 50)

    before = count_sanctioned()
    print(f"Addresses in DB before update: {before}")

    try:
        count = await update_sanctions_list()
        after = count_sanctioned()
        print(f"Addresses in DB after update:  {after}")
        print(f"✅ Done — {count} ETH addresses loaded from OFAC SDN list.")
    except Exception as e:
        print(f"❌ Update failed: {e}")
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())

File: scripts/update_sanctions.ts
import fs from "fs";
import path from "path";
import { updateSanctionsList, countSanctioned } from "../src/lib/forensics/sanctions";

// Custom helper to load .env variables
function loadEnv() {
  const envPath = path.join(process.cwd(), ".env");
  if (fs.existsSync(envPath)) {
    fs.readFileSync(envPath, "utf-8")
      .split("\n")
      .forEach((line) => {
        const match = line.match(/^\s*([\w.-]+)\s*=\s*(.*)?\s*$/);
        if (match) {
          const key = match[1];
          let val = match[2] || "";
          if (val.startsWith('"') && val.endsWith('"')) val = val.slice(1, -1);
          else if (val.startsWith("'") && val.endsWith("'")) val = val.slice(1, -1);
          process.env[key] = val;
        }
      });
  }
}
loadEnv();

async function main() {
  console.log("Chain Analytics — OFAC Sanctions Update");
  console.log("=".repeat(50));

  const before = await countSanctioned();
  console.log(`Addresses in DB before update: ${before}`);

  try {
    const count = await updateSanctionsList();
    const after = await countSanctioned();
    console.log(`Addresses in DB after update:  ${after}`);
    console.log(`✅ Done — ${count} ETH addresses loaded from OFAC SDN list.`);
    process.exit(0);
  } catch (e: any) {
    console.error(`❌ Update failed: ${e.message || e}`);
    process.exit(1);
  }
}

main();
